velox/common/memory/HashStringAllocator.h

```diff
@@ -328,20 +358,35 @@ class HashStringAllocator : public StreamArena {
   // return a block that is much larger than preferredSize. Otherwise,
   // the block can be larger and the user is expected to call
   // freeRestOfBlock to finalize the allocation.
-  Header* FOLLY_NULLABLE allocateFromFreeList(
+  Header* FOLLY_NULLABLE allocateFromFreeLists(
       int32_t preferredSize,
       bool mustHaveSize,
       bool isFinalSize);
 
+  Header* FOLLY_NULLABLE allocateFromFreeList(
+      int32_t preferredSize,
+      bool mustHaveSize,
+      bool isFinalSize,
+      int32_t freeListIndex);
+
   // Sets 'header' to be 'keepBytes' long and adds the remainder of
   // 'header's memory to free list. Does nothing if the resulting
   // blocks would be below minimum size.
   void freeRestOfBlock(Header* FOLLY_NONNULL header, int32_t keepBytes);
 
+  // Returns the free list index for 'size'. Masks out empty sizes that can be
+  // given by 'mask'. If 'mask' excludes all free lists, returns >
+  // kNumFreeLists.
+  int32_t freeListIndex(int32_t size, uint32_t mask = ~0);
+
   // Circular list of free blocks.
-  CompactDoubleList free_;
+  CompactDoubleList free_[kNumFreeLists];
+
+  // Bitmap with a 1 if the corresponding list in 'free_' is not empty.
+  int32_t freeHasData_{0};
```

Comment:
The name freeHasData is confusing because "free" block is supposed to not have any data. I wonder if there is a better name.